ui_framework/src/components/table/listing_table/listing_table.js

```diff
@@ -0,0 +1,165 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import _ from 'lodash';
+
+import { KuiListingTableToolBar } from './listing_table_tool_bar';
+import { KuiListingTableToolBarFooter } from './listing_table_tool_bar_footer';
+import { KuiListingTableRow } from './listing_table_row';
+
+import {
+  KuiControlledTable,
+  KuiTableHeaderCheckBoxCell,
+  KuiTableBody,
+  KuiTableHeader,
+  KuiTable,
+  KuiTableHeaderCell,
+} from '../../index';
+
+import {
+  LEFT_ALIGNMENT,
+  RIGHT_ALIGNMENT,
+} from '../../../services';
+
+export function KuiListingTable({
+  rows,
+  header,
+  pager,
+  toolBarActions,
+  onFilter,
+  onItemSelectionChanged,
+  selectedRowIds,
+  filter,
+  prompt,
+}) {
+
+  function areAllRowsSelected() {
+    return rows.length === selectedRowIds.length;
+  }
+
+  function toggleAll() {
+    if (areAllRowsSelected()) {
+      onItemSelectionChanged([]);
+    } else {
+      onItemSelectionChanged(rows.map(row => row.id));
+    }
+  }
+
+  function toggleRow(rowId) {
+    const selectedRowIndex = selectedRowIds.indexOf(rowId);
+    if (selectedRowIndex >= 0) {
+      onItemSelectionChanged(selectedRowIds.filter((item, index) => index !== selectedRowIndex));
+    } else {
+      onItemSelectionChanged([
+        ...selectedRowIds,
+        rowId
+      ]);
+    }
+  }
+
+  function renderTableRows() {
+    return rows.map((row, rowIndex) => {
+      return (
+        <KuiListingTableRow
+          key={rowIndex}
+          isSelected={selectedRowIds.indexOf(row.id) >= 0}
+          onSelectionChanged={toggleRow}
+          row={row}
+        />
+      );
+    });
+  }
+
+  function renderHeader() {
+    return header.map((cell, index) => {
+      let { content, ...props } = cell;
+      if (React.isValidElement(cell) || !_.isObject(cell)) {
+        props = [];
+        content = cell;
+      }
+      return (
+        <KuiTableHeaderCell
+          key={index}
+          {...props}
+        >
+          {content}
+        </KuiTableHeaderCell>
+      );
+    });
+  }
+
+  function renderInnerTable() {
+    return (
+      <KuiTable>
+        <KuiTableHeader>
+          <KuiTableHeaderCheckBoxCell
+            isChecked={areAllRowsSelected()}
+            onChange={toggleAll}
+          />
+          {renderHeader()}
+        </KuiTableHeader>
+
+        <KuiTableBody>
+          {renderTableRows()}
+        </KuiTableBody>
+      </KuiTable>
+    );
+  }
+
+  return (
+    <KuiControlledTable>
+      <KuiListingTableToolBar
+        actions={toolBarActions}
+        pager={pager}
+        onFilter={onFilter}
+        filter={filter}
+      />
+
+      {prompt ? prompt : renderInnerTable()}
+
+      <KuiListingTableToolBarFooter
+        itemsSelectedCount={selectedRowIds.length}
+        pager={pager}
+      />
+    </KuiControlledTable>
+  );
+}
+
+KuiListingTable.PropTypes = {
+  header: PropTypes.arrayOf(
```

Comment:
What's the thinking behind this being data drive, instead of component driven? Like why not construct some component in the caller and pass it in directly?

This is what is passed in now:
```
return [
      {
        content: 'Title',
        onSort: this.sortByTitle,
        isSorted: this.state.sortedColumn === 'title',
        isSortAscending: this.sortableProperties.isAscendingByName('title'),
      },
      {
        content: 'Type',
        onSort: this.sortByType,
        isSorted: this.state.sortedColumn === 'type',
        isSortAscending: this.sortableProperties.isAscendingByName('type'),
      },
    ];
```

But what if we just did:
```
return [
  <KuiTableHeaderCell
       onSort: this.sortByTitle,
        isSorted: this.state.sortedColumn === 'title',
        isSortAscending: this.sortableProperties.isAscendingByName('title')
   >
      Title
   </KuiTableHeaderCell>
];
```

or something similar?

It feels we have too much unnecessary code around this header concept and a simpler solution might work out better. Thoughts?

Reply:
This same question/concept applies to maybe some of the other props here, like rows.

I think I understand the inclination to separate the data from the presentational concerns/details, but it feels the coupling is pretty strong already so I'm not sure it's buying us much.

Reply:
It was that way at first.  It changed in this commit- https://github.com/elastic/kibana/pull/14227/commits/f50f916d536370e5adf313ef8f79113dc9c53b19
after a discussion with @cjcenizal.  I like that it's more refactored in the data driven approach.  Comment that initiated the change is here: https://github.com/elastic/kibana/pull/14227/commits/f50f916d536370e5adf313ef8f79113dc9c53b19#r142833708

Reply:
Yea that makes sense. I'm not sure I love it still, but it's perfectly readable and still easy to understand so carry on!

Reply:
For context, I've copy/pasted my original comment below. I was hoping that by encapsulating the row, header, and cell components within the table component, it would let the consumer use a simpler interface to specify their content. For example, if the rows' cells just needs to contain text, you can specify each row as an array of strings instead of needing to specify each cell component too. I was also hoping that that this would move some of the sorting logic into KuiListingTable.

That said, I'm not 100% sold on this approach and really welcome people's thoughts on it. I'd love it if we could improve these abstractions, but I'm also OK with reverting it if people find them too difficult to use.

---

I wonder if there's a way where we can create abstractions for the way headers and rows present data. If the interfaces of these abstractions were to be similar then I think that would be ergonomic for the consumer, since they're very similar components (they're both containers of the same number of cells).

I'm thinking of something like this:

```jsx
// This component is responsible for instantiating `KuiTableRowCell` instances.
// It will iterate over each element in the cells array. You can specify a cell as a string,
// an element, a number, or a more complex configuration object. Depending
// on how the type is formatted, the values will either be passed as `children`
// or as various props to each `KuiTableRowCell`. 

const rows = this.items.map(item => (
  <KuiListingTableRow
    cells={[
      <a className="kuiLink" href={item.link}>item.name</a>,
      <div className={`kuiIcon kuiIcon--success ${item.icon}`}/>,
      item.date,
      {
        content: item.magnitude,
        align: 'right',
      },
    ]}
  />
));

// Same idea here.

const header = <KuiListingTableHeader
  cells={[{
    {
      content: 'Name',
      sortableId: 'name, // The presence of this property will present a column as sortable
    },
    'Icon',
    {
      content: 'Date',
      sortableId: 'date,
    }, {
      content: 'Magnitude',
      align: 'right',
      sortableId: 'magnitude',
    }
  }]}
/>

// And then maybe we could leverage `children' and the table would become this:

<KuiListingTable
  pager={this.renderPager()}
  toolbarActions={this.renderToolbarActions()}
  selectedRowIds={this.state.selectedRowIds}
  header={header}
  onFilter={() => {}}
  filter=""
  onItemSelectionChanged={this.onItemSelectionChanged}
>
  {rows}
</KuiListingTable>
```

What do you think? Does this sound crazy?

Reply:
Right, as @cjcenizal said, you can still pass an array of nodes or strings for the header and rows, it's just the internal part, not to include the KuiTableRowCell part.  But this means we have to pass in sort information and align information sometimes, for the props that go on that outer element.

I'm also flexible if we find a better approach.  For now I'll go ahead and submit once this passes, and we can update when/if we find some better abstractions or patterns.  Perhaps once we add in an HOC and refactor the sorting logic, we'll find a better way to handle it.

Reply:
It's easy to follow and totally makes sense. I can't articulate my thoughts against it that well, which could easily indicate they aren't grounded in anything useful. Please continue as is and I'm sure we'll find another time to rethink it and maybe have some more useful conversations (on my side at least!)